python/mp/rosary.py

```diff
@@ -400,56 +553,104 @@ def mainloop(self):
-            self.run_mainloop = False
+            # Start fading out all existing effects
+            self.clear_effects_fade()
 
-    def generate_osc_path(self, fn_name):
+            self.add_trigger_object(requested_trigger(*args, **kwargs))
+
+
+    def turn_knob(self, knob_name, *args, **kwargs):
         """
-        I know this is not consistent with the pattern already established
-        in effect.py, but I feel like just calling /rosary is more intuitive
-        ...for now
+        For all registered effect instance methods mapped to this knob_name,
+        turn the knob (we're assured they're all running effects)
         """
 
-        return "/rosary/{}".format(fn_name)
+        for fn, instance in self.knobs.get(knob_name, []):
+            fn(instance, *args, **kwargs)
+
 
-    def register_with_dispatcher(self):
+    def route_osc_call(self, full_path, *args, **kwargs):
         """
-        Maybe I could move this function to the DispatcherMapper class.
-        Mabye another day.
+        Figure out what the OSC path means and, well, do what the
+        runes instruct us to do.
         """
 
-        print("* ROSARY PATHS *")
-        for fn_name in self.dm.registered_methods.keys():
+        osc_args = full_path.split('/')
+        osc_args.remove('')
+
+        # 'rosary', 'effect', 'trigger', etc.
+        namespace = osc_args.pop(0)
+        # The actual function name, in that namespace: 'add_effect', etc.
+        fn_name = osc_args.pop(0)
+
+        # Finally, the rest of the parsed args are expected to actually
+        # represent key/value pairs, so expect there to be an even-numbered
+        # count of items
+        inferred_kwargs = {}
+        for i in range(int(len(osc_args)/2)):
+            implied_arg = osc_args[i*2]
+            implied_val = osc_args[i*2 + 1]
+
+            # Convert strings representing ints to ints and
+            # strings representing floats to floats
+            try:
+                implied_val = ast.literal_eval(implied_val)
+            except ValueError:
+                pass
+
+            inferred_kwargs[implied_arg] = implied_val
+
+        # If we have inferred kwargs, use them
+        # Otherwise, use whatever is in passed args
+        # If nothing, assume a 1 from passed args
+        if namespace == 'rosary':
+            if inferred_kwargs:
+                # We expect to mostly end up here
+                if fn_name in self.dm.exposed_methods.keys():
+                    self.dm.exposed_methods[fn_name](self, **inferred_kwargs)
+                # But in some rare cases, we set some reasonable defaults
+                # even if you're being COMPLETELY lazy
+                else:
+                    self.dm.exposed_methods[fn_name](self, *args)
+
+        elif namespace == 'effect':
+            if inferred_kwargs:
+                self.turn_knob(fn_name, **inferred_kwargs)
+            else:
+                self.turn_knob(fn_name, *args)
```

Comment:
Any `/effect` paths try to find a `@dm.expose()`'d "knob" in the individual effects: e.g. bounce.py, sine_wave.py.